src/ctelnet.cpp

```diff
@@ -1023,86 +1099,112 @@ void cTelnet::processTelnetCommand(const string& command)
+            // bytes from beginning (3) and end (2):
+            payload = payload.mid(3, static_cast<int>(payload.size()) - 5);
+
             // strip first 3 characters to get rid of <IAC><SB><201>
             // and strip the last 2 characters to get rid of <IAC><TN_SE>
-            setGMCPVariables(rawPayload.mid(3, rawPayload.size() - 5));
+            setGMCPVariables(payload);
             return;
         }
 
-        if (option == static_cast<unsigned char>(102)) {
-            QString _m = command.c_str();
-            if (command.size() < 6) {
+        if (option == OPT_102) {
+            QByteArray payload = command.c_str();
+            if (payload.size() < 6) {
                 return;
             }
-            _m = _m.mid(3, command.size() - 5);
-            setChannel102Variables(_m);
+            // payload is in the Mud Server's encoding, trim off the Telnet suboption
+            // bytes from beginning (3) and end (2):
+            payload = payload.mid(3, static_cast<int>(payload.size()) - 5);
+
+            setChannel102Variables(payload);
             return;
         }
-        switch (option) //switch 2
-        {
+
+        switch (option) { //switch 2
         case OPT_STATUS: {
-            //see OPT_TERMINAL_TYPE for explanation why I'm doing this
-            if (true) {
-                qDebug() << "WARNING: FIXME #501";
-                if (command[3] == TNSB_SEND) {
-                    qDebug() << "WARNING: FIXME #504";
-                    //request to send all enabled commands; if server sends his
-                    //own list of commands, we just ignore it (well, he shouldn't
-                    //send anything, as we do not request anything, but there are
-                    //so many servers out there, that you can never be sure...)
-                    string cmd;
-                    cmd += TN_IAC;
-                    cmd += TN_SB;
-                    cmd += OPT_STATUS;
-                    cmd += TNSB_IS;
-                    for (short i = 0; i < 256; i++) {
-                        if (myOptionState[i]) {
-                            cmd += TN_WILL;
+            if (command.length() >= 6 && command[3] == TNSB_SEND && command[4] == TN_IAC && command[5] == TN_SE) {
+                //request to send all enabled commands; if server sends his
+                //own list of commands, we just ignore it (well, he shouldn't
+                //send anything, as we do not request anything, but there are
+                //so many servers out there, that you can never be sure...)
+                // FIXME: This is damaged at the moment as we do not properly take care of the bits for the protocols that we manage ourselves e.g. ATCP/GMCP/MSDP/MXP etc...
+                string cmd;
+                cmd += TN_IAC;
+                cmd += TN_SB;
+                cmd += OPT_STATUS;
+                cmd += TNSB_IS;
+                for (size_t i = 0; i < 256; ++i) {
```

Comment:
Why the datatype change?

Reply:
`i` is an index into a C array - it is ***supposed*** to be done with a `size_t` type!